hdat/hdat_cli.py

```diff
@@ -0,0 +1,113 @@
+from .runner import run_cases
+from .util import AbortError
+
+
+def parse_arguments(arguments):
+    parser = argparse.ArgumentParser(prog='hdat')
+    subparsers = parser.add_subparsers(dest='command', metavar='<command>')
+
+    run_help = 'run cases, store results in archive, compare against goldens'
+    run_parser = subparsers.add_parser('run', help=run_help)
+    run_parser.add_argument('casespecs', nargs='*', default=[''], metavar='<case>')
+    run_parser.add_argument('--collect-only', default=False, action='store_true')
+
+    show_help = 'visualize a single result'
+    show_parser = subparsers.add_parser('show', help=show_help)
+    show_result_help = 'result specifier to show'
+    show_parser.add_argument('resultspec', nargs="?", default='', metavar='<result>', help=show_result_help)
+
+    runshow_help = 'run then visualize a single result'
+    runshow_parser = subparsers.add_parser('runshow', help=runshow_help)
+    runshow_parser.add_argument('casespecs', nargs=1, default='', metavar='<result>')
+
+    diff_help = 'compare two results'
+    diff_parser = subparsers.add_parser('diff', help=diff_help)
+    diff_golden_help = 'result to compare to (defaults to current golden for the case)'
+    diff_parser.add_argument('goldenspec', nargs='?', default=None, metavar='<golden>', help=diff_golden_help)
+    diff_result_help = 'result being compared'
+    diff_parser.add_argument('resultspec', nargs=1, metavar='<result>', help=diff_result_help)
+
+    verify_help = 'move result metrics from archive to the golden store'
+    verify_parser = subparsers.add_parser('verify', help=verify_help)
+    verify_result_help = 'results to be stripped and moved into the golden store'
+    verify_parser.add_argument('resultspec', nargs='?', default='', metavar='<result>', help=verify_result_help)
+
+    return parser.parse_args(arguments)
+
+
+def _format_cases_status(cases_status):
+    return 'PASS: {}, FAIL: {}, UNKNOWN: {}, ERROR: {}'.format(
+        cases_status['pass'],
+        cases_status['fail'],
+        cases_status['unknown'],
+        cases_status['error'],
+    )
+
+
+def hdat_cli(arguments, suites, golden_store, archive, git_info):
+    args = parse_arguments(arguments)
+
+    if args.command is None:
+        parse_arguments(['-h'])
+
+    if args.command == 'run' and args.collect_only:
+        cases = resolve_casespecs(suites, args.casespecs)
+        print("\n".join(['{}/{}'.format(s, c) for s, c in cases]))
+    elif args.command == 'run':
+        cases = resolve_casespecs(suites, args.casespecs)
+        cases_status = run_cases(suites, golden_store, archive, git_info, cases)
+        if cases_status['pass'] < len(cases):
+            raise AbortError(_format_cases_status(cases_status))
+    elif args.command == 'show':
+        results = resolve_resultspecs(archive, args.resultspec)
+        for result in results:
+            show_result(suites, result)
+    elif args.command == 'runshow':
```

Comment:
When I added the `runshow` command, that I couldn't think of as simple way to avoid duplication between `run`, `show`, and `runshow`.  That said, I suspect there is a good way to do it.

Reply:
The duplication isn't so bad, but the commands could be broken out from the if/elif blocks into their own functions.